### PR TITLE
Simplify the break rule of logical optimizer passes

### LINKED ISSUE #4841: Optimize the break rule of optimizer passes.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
https://github.com/apache/arrow-datafusion/blob/2db3d2eae9f4e51f01cb705677c8c231b43a6172/datafusion/optimizer/src/optimizer.rs#L312-L320

We don't need the string comparison anymore because we have had the `try_optimize` method in the `OptimizerRule` which returns `None` when the plan is not optimized.

**Describe the solution you'd like**
Simplify the above logic.

**Describe alternatives you've considered**
We could not do this if there is any reason.

**Additional context**

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Which issue does this PR close?

Closes #4841.

# Rationale for this change

# What changes are included in this PR?

Remove the string comparison logical.
Add a flag `plan_is_changed` to track whether the plan is optimized in one optimizer pass. 

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jackwener** [comment]
When `result` is `Ok(Some())`, it doesn't means that plan have changed.

Sometimes, plan didn't changed, some rule also will return `Ok(Some())`.

File: datafusion/optimizer/src/optimizer.rs [hunk 1]
```diff
@@ -309,17 +309,11 @@ impl Optimizer {
             }
             log_plan(&format!("Optimized plan (pass {i})"), &new_plan);
 
-            // TODO this is an expensive way to see if the optimizer did anything and
-            // it would be better to change the OptimizerRule trait to return an Option
-            // instead
-            let new_plan_str = format!("{}", new_plan.display_indent());
-            if plan_str == new_plan_str {
+            if !plan_is_changed {
```
Comment (alamb) [inline datafusion/optimizer/src/optimizer.rs:312]: I agree with  @jackwener that that returning `Ok(Some(plan))` may not actually have changed the plan

Perhaps we need to have a `impl PartialEq` for `LogicalPlan` that does the correct deep equality check?

https://github.com/apache/arrow-datafusion/blob/3cc607de4ce6e9e1fd537091e471858c62f58653/datafusion/expr/src/logical_plan/plan.rs#L50

I remember at some point we tried to `#[derive(PartialEq)]` for `LogicalPlan` but we hit some problem 🤔  maybe it has been resolved. Or maybe we can manually derive `PartialEq` 🤔

**Review by alamb — COMMENTED** [review]
Thank you @HaoYang670

**Comment by HaoYang670** [comment]
Thank you for your review @jackwener and @alamb 😁.

> Sometimes, plan didn't changed, some rule also will return Ok(Some()).
e.g: some rule recurse inside themselves, and alway return Ok(Some())

Hi @jackwener, is this the expected behavior or a known bug that hasn't been fixed yet?

**Comment by HaoYang670** [comment]
>e.g: some rule recurse inside themselves, and alway return Ok(Some())
It's expected behavior.

Why don't we return None if the rule can't optimize the plan is this scenario?

**Comment by jackwener** [comment]
> > e.g: some rule recurse inside themselves, and alway return Ok(Some())
> > It's expected behavior.
> 
> Why don't we return None if the rule can't optimize the plan is this scenario?

Some rule I modify them to avoid self-recurse, those rules is ok. But other rules still use original code, always return Ok(Some())

**Comment by alamb** [comment]
> Why don't we return None if the rule can't optimize the plan is this scenario?

I don't think there is a theoretical reason we don't do this (and it would probably be more efficient) but I don't think it is how the code works today. I would definitely be supportive of updating the passes so they had this property, however 👍

**Comment by alamb** [comment]
Closing as this PR is over a year old. Please feel free to reopen it / rebase it if you plan to keep working on it

**Pull request closed without merging**